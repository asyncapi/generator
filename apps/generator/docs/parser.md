---
title: "Parser"
weight: 60
---

## Parser

The **AsyncAPI Parser** is a powerful library for parsing and validating [AsyncAPI documents](asyncapi-document) in both Node.js and browser environments. It supports YAML and JSON inputs and ensures correctness through schema validation.

### âœ… Supported schema formats:

- AsyncAPI (v2 and v3)
- OpenAPI
- JSON Schema
- Avro
- RAML

These schemas are automatically transformed into JSON Schema format internally, simplifying access and manipulation.

---

## Key Features

- **Parsing** of AsyncAPI YAML/JSON files.
- **Validation** against AsyncAPI and referenced schemas.
- **Dereferencing** of `$ref` values into inlined content.
- **Helper functions** to access channels, messages, schemas, and more.

---

## Example: AsyncAPI v3 Document

Below is a minimal v3.0.0-compliant AsyncAPI document using `channels`, `operations`, and component references:

```yaml
asyncapi: 3.0.0
info:
  title: Comments Service
  version: 1.0.0
  description: Handles all comment-related events.
servers:
  dev:
    host: test.mosquitto.org
    protocol: mqtt
channels:
  sendCommentLiked:
    address: comment/liked
    messages:
      commentLiked:
        $ref: '#/components/messages/commentLiked'
  sendCommentUnliked:
    address: comment/unliked
    messages:
      commentUnliked:
        $ref: '#/components/messages/commentUnliked'
operations:
  sendCommentLiked:
    action: send
    channel:
      $ref: '#/channels/sendCommentLiked'
  sendCommentUnliked:
    action: send
    channel:
      $ref: '#/channels/sendCommentUnliked'
components:
  messages:
    commentLiked:
      payload:
        $ref: '#/components/schemas/commentReaction'
    commentUnliked:
      payload:
        $ref: '#/components/schemas/commentReaction'
  schemas:
    commentReaction:
      type: object
      properties:
        commentId:
          type: string
