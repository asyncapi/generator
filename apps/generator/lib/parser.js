const fs = require('fs');
const { convertToOldAPI } = require('@asyncapi/parser');
const { ConvertDocumentParserAPIVersion, NewParser } = require('@smoya/multi-parser');

const parser = module.exports;

/**
 * Convert the template defined value `apiVersion: 'v1'` to only contain the numeric value `1`.
 */
parser.sanitizeTemplateApiVersion = (apiVersion) => {
  if (!apiVersion) return;
  if (apiVersion && apiVersion.length > 1) {
    return Number(apiVersion.substring(1));
  }
  return Number(apiVersion);
};

parser.parse = async (asyncapi, oldOptions, generator) => {
  let apiVersion = this.sanitizeTemplateApiVersion(generator.templateConfig.apiVersion);
  // Defaulting to apiVersion v1 to convert it to the Parser-API v1 afterwards.
  if (!this.usesNewAPI(generator.templateConfig)) {
    apiVersion = 1;
  }
  const options = convertOldOptionsToNew(oldOptions, generator);
  const parser = NewParser(apiVersion, {parserOptions: options, includeSchemaParsers: true});
  const { document, diagnostics } = await parser.parse(asyncapi, options);
  if (!document) {
    return {document, diagnostics};
  }
  const correctDocument = this.getProperApiDocument(document, generator.templateConfig);
  return {document: correctDocument, diagnostics};
};

/**
 * If the template expect one of the Parser-API versions, it must be above 0
 */
parser.usesNewAPI = (templateConfig = {}) => {
  return this.sanitizeTemplateApiVersion(templateConfig.apiVersion) > 0;
};

/**
 * Based on the current parsed AsyncAPI document, convert it to expected API version from the template.
 */
parser.getProperApiDocument = (asyncapiDocument, templateConfig = {}) => {
  const apiVersion = this.sanitizeTemplateApiVersion(templateConfig.apiVersion);
  if (apiVersion === undefined) {
    // Convert to old API from JS Parser v1
    return convertToOldAPI(asyncapiDocument);
  }
  return ConvertDocumentParserAPIVersion(asyncapiDocument, apiVersion);
};

// The new options for the v2 Parser are different from those for the v1 version, but in order not to release Generator v2, we are converting the old options of Parser to the new ones.
function convertOldOptionsToNew(oldOptions, generator) {
  if (!oldOptions) return;
  const newOptions = {};

  if (typeof oldOptions.path === 'string') {
    newOptions.source = oldOptions.path;
  }
  if (typeof oldOptions.applyTraits === 'boolean') {
    newOptions.applyTraits = oldOptions.applyTraits;
  }

  const resolvers = [];
  if (generator && generator.mapBaseUrlToFolder && generator.mapBaseUrlToFolder.url) {
    resolvers.push(...getMapBaseUrlToFolderResolvers(generator.mapBaseUrlToFolder));
  }
  if (oldOptions.resolve) {
    resolvers.push(...convertOldResolvers(oldOptions.resolve));
  }

  if (resolvers.length) {
    newOptions.__unstable = {};
    newOptions.__unstable.resolver = {
      resolvers,
    };
  }

  return newOptions;
}

/**
 * Creates a custom resolver that maps urlToFolder.url to urlToFolder.folder
 * Building your custom resolver is explained here: https://apitools.dev/json-schema-ref-parser/docs/plugins/resolvers.html
 *
 * @private
 * @param  {object} urlToFolder to resolve url e.g. https://schema.example.com/crm/ to a folder e.g. ./test/docs/.
 * @return {{read(*, *, *): Promise<unknown>, canRead(*): boolean, order: number}}
 */
function getMapBaseUrlToFolderResolvers({ url: baseUrl, folder: baseDir }) {
  const resolver = {
    order: 1,
    canRead: true,
    read(uri) {
      return new Promise(((resolve, reject) => {
        const path = uri.toString();
        const localpath = path.replace(baseUrl, baseDir);
        try {
          fs.readFile(localpath, (err, data) => {
            if (err) {
              reject(`Error opening file "${localpath}"`);
            } else {
              resolve(data.toString());
            }
          });
        } catch (err) {
          reject(`Error opening file "${localpath}"`);
        }
      }));
    }
  };

  return [
    { schema: 'http', ...resolver, },
    { schema: 'https', ...resolver, },
  ];
};

function convertOldResolvers(resolvers = {}) { // NOSONAR
  if (Object.keys(resolvers).length === 0) return [];

  return Object.entries(resolvers).map(([protocol, resolver]) => {
    return {
      schema: protocol,
      order: resolver.order || 1,
      canRead: (uri) => canReadFn(uri, resolver.canRead),
      read: (uri) => {
        return resolver.read({ url: uri.valueOf(), extension: uri.suffix() });
      }
    };
  });
}

function canReadFn(uri, canRead) {
  const value = uri.valueOf();
  if (typeof canRead === 'boolean') return canRead;
  if (typeof canRead === 'string') return canRead === value;
  if (Array.isArray(canRead)) return canRead.includes(value);
  if (canRead instanceof RegExp) return canRead.test(value);
  if (typeof canRead === 'function') {
    return canRead({ url: value, extension: uri.suffix() });
  }
  return false;
}
