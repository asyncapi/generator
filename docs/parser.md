---
title: "Parser"
weight: 90
---

>This document is under construction.