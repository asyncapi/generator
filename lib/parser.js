const fs = require('fs');

const { Parser, convertToOldAPI } = require('@asyncapi/parser/cjs');
const { OpenAPISchemaParser } = require('@asyncapi/openapi-schema-parser');
const { AvroSchemaParser } = require('@asyncapi/avro-schema-parser');
const { RamlDTSchemaParser } = require('@asyncapi/raml-dt-schema-parser');
const { ProtoBuffSchemaParser } = require('@asyncapi/protobuf-schema-parser');

const parser = module.exports;

const defaultParser = new Parser({
  schemaParsers: [
    OpenAPISchemaParser(),
    AvroSchemaParser(),
    RamlDTSchemaParser(),
    ProtoBuffSchemaParser(),
  ],
});

parser.parse = (asyncapi, oldOptions, generator) => {
  const options = convertOldOptionsToNew(oldOptions, generator);
  return defaultParser.parse(asyncapi, options);
};

parser.usesNewAPI = (templateConfig = {}) => {
  return templateConfig.apiVersion === 'v1';
};

parser.getProperApiDocument = (asyncapiDocument, templateConfig) => {
  return parser.usesNewAPI(templateConfig) ? asyncapiDocument : convertToOldAPI(asyncapiDocument);
};

// The new options for the v2 Parser are different from those for the v1 version, but in order not to release Generator v2, we are converting the old options of Parser to the new ones.
function convertOldOptionsToNew(oldOptions, generator) {
  if (!oldOptions) return;
  const newOptions = {};

  if (typeof oldOptions.path === 'string') {
    newOptions.source = oldOptions.path;
  }
  if (typeof oldOptions.applyTraits === 'boolean') {
    newOptions.applyTraits = oldOptions.applyTraits;
  }

  const resolvers = [];
  if (generator && generator.mapBaseUrlToFolder && generator.mapBaseUrlToFolder.url) {
    resolvers.push(...getMapBaseUrlToFolderResolvers(generator.mapBaseUrlToFolder));
  }
  if (oldOptions.resolve) {
    resolvers.push(...convertOldResolvers(oldOptions.resolve));
  }

  if (resolvers.length) {
    newOptions.__unstable = {};
    newOptions.__unstable.resolver = {
      resolvers,
    };
  }

  return newOptions;
}

/**
 * Creates a custom resolver that maps urlToFolder.url to urlToFolder.folder
 * Building your custom resolver is explained here: https://apitools.dev/json-schema-ref-parser/docs/plugins/resolvers.html
 *
 * @private
 * @param  {object} urlToFolder to resolve url e.g. https://schema.example.com/crm/ to a folder e.g. ./test/docs/.
 * @return {{read(*, *, *): Promise<unknown>, canRead(*): boolean, order: number}}
 */
function getMapBaseUrlToFolderResolvers({ url: baseUrl, folder: baseDir }) {
  const resolver = {
    order: 1,
    canRead: true,
    read(uri) {
      return new Promise(((resolve, reject) => {
        const path = uri.toString();
        const localpath = path.replace(baseUrl, baseDir);
        try {
          fs.readFile(localpath, (err, data) => {
            if (err) {
              reject(`Error opening file "${localpath}"`);
            } else {
              resolve(data.toString());
            }
          });
        } catch (err) {
          reject(`Error opening file "${localpath}"`);
        }
      }));
    }
  };

  return [
    { schema: 'http', ...resolver, },
    { schema: 'https', ...resolver, },
  ];
};

function convertOldResolvers(resolvers = {}) { // NOSONAR
  if (Object.keys(resolvers).length === 0) return [];

  return Object.entries(resolvers).map(([protocol, resolver]) => {
    return {
      schema: protocol,
      order: resolver.order || 1,
      canRead: (uri) => canReadFn(uri, resolver.canRead),
      read: (uri) => {
        return resolver.read({ url: uri.valueOf(), extension: uri.suffix() });
      }
    };
  });
}

function canReadFn(uri, canRead) {
  const value = uri.valueOf();
  if (typeof canRead === 'boolean') return canRead;
  if (typeof canRead === 'string') return canRead === value;
  if (Array.isArray(canRead)) return canRead.includes(value);
  if (canRead instanceof RegExp) return canRead.test(value);
  if (typeof canRead === 'function') {
    return canRead({ url: value, extension: uri.suffix() });
  }
  return false;
}
