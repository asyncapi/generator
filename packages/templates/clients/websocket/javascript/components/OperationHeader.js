import { Text } from '@asyncapi/generator-react-sdk';

export default function OperationHeader({operation}) {
  const operationId = operation.id();
  return (
    <Text>
      {`#### \`${operationId}(payload)\`
${operation.hasSummary() ?  operation.summary() : ''}
${operation.hasDescription() ? `\n${operation.description()}` : ''}`}
    </Text>
  );
}