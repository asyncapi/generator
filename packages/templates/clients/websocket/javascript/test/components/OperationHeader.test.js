import path from 'path';
import { render } from '@asyncapi/generator-react-sdk';
import { Parser, fromFile } from '@asyncapi/parser';
import OperationHeader from '../../components/OperationHeader';

const parser = new Parser();
const asyncapi_websocket_query = path.resolve(__dirname, '../../../../../../helpers/test/__fixtures__/asyncapi-websocket-query.yml');

describe('Testing of OperationHeader function', () => {
  let parsedAsyncAPIDocument;

  beforeAll(async () => {
    const parseResult = await fromFile(parser, asyncapi_websocket_query).parse();
    parsedAsyncAPIDocument = parseResult.document;
  });

  test('render operation header with summary and description correctly', () => {
    const operation = parsedAsyncAPIDocument.operations().get('mixedMessageExamples');
    const result = render(<OperationHeader operation={operation} />);
    const actual = result.trim();
    expect(actual).toMatchSnapshot();
  });

  test('render operation header with summary only correctly', () => {
    const operation = parsedAsyncAPIDocument.operations().get('noMessage');
    const result = render(<OperationHeader operation={operation} />);
    const actual = result.trim();
    expect(actual).toMatchSnapshot();
  });

  test('render operation header with description only correctly', () => {
    const operation = parsedAsyncAPIDocument.operations().get('noSummaryOperations');
    const result = render(<OperationHeader operation={operation} />);
    const actual = result.trim();
    expect(actual).toMatchSnapshot();
  });

  test('render operation header with no summary or description correctly', () => {
    const operation = parsedAsyncAPIDocument.operations().get('noSummaryNoDescriptionOperations');
    const result = render(<OperationHeader operation={operation} />);
    const actual = result.trim();
    expect(actual).toMatchSnapshot();
  });
});