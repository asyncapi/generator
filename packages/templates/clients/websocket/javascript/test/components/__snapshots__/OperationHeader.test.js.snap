// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing of OperationHeader function render operation header with description only correctly 1`] = `
"#### \`noSummaryOperations(payload)\`


This is a description without a summary"
`;

exports[`Testing of OperationHeader function render operation header with no summary or description correctly 1`] = `"#### \`noSummaryNoDescriptionOperations(payload)\`"`;

exports[`Testing of OperationHeader function render operation header with summary and description correctly 1`] = `
"#### \`mixedMessageExamples(payload)\`
Mixed message example coverage

An example of a send operation with multiple message with examples"
`;

exports[`Testing of OperationHeader function render operation header with summary only correctly 1`] = `
"#### \`noMessage(payload)\`
Operation with no messages"
`;
